src/text/non_blocking/raw_text_reader.rs

```diff
@@ -207,6 +227,67 @@ impl<A: AsRef<[u8]>> RawTextReader<A> {
         Ok(())
     }
 
+    // This function implements the logic of `step_out`, but is extracted in order to make tracking
+    // of the step_out depth, which is handled by the actual `step_out` function, easier.
+    //
+    // See `step_out` for more info about how we handle errors when stepping out.
+    fn step_out_impl(&mut self) -> IonResult<()> {
+        // If we do not yet have a target depth, establish it.
+        if self.step_out_depth.is_none() {
+            self.step_out_depth = Some(self.depth() - 1);
+        }
+
+        if self.carried_state != ReaderState::ContainerDelimiter {
+            if self.parents.is_empty() {
+                return illegal_operation(
+                    "Cannot call `step_out()` when the reader is at the top level.",
+                );
+            }
+
+            // The container we're stepping out of.
+            let parent = self.parents.last().unwrap();
+
+            // If we're not at the end of the current container, advance the cursor until we are.
+            // Unlike the binary reader, which can skip-scan, the text reader must visit every value
+            // between its current position and the end of the container.
+            if !parent.is_exhausted() {
+                while let RawStreamItem::Value(_) | RawStreamItem::Null(_) = self.next()? {}
+            }
+
+            // Remove the parent container from the stack and clear the current value.
+            let _ = self.parents.pop();
+            self.current_value = None;
+
+            if self.parents.is_empty() {
+                // We're at the top level; nothing left to do.
+                return Ok(());
+            }
+            // We have stepped out and need to track that we have in just in case we error due to
+            // incomplete text while looking.
+            self.carried_state = ReaderState::ContainerDelimiter;
```

Comment:
The `ContainerDelimiter` state is specifically for when we have finished reading all of the container's items and have stepped out of the container, but need to find an inner delimiter, or closing delimiter, for a parent container. That is why it is set after exhausting the contents, and after we test for an empty `parents` list.

If an error occurs while reading the container's values, like your example, the carried state *needs* to be the state from where the error occurred. Since the calls to `next` in the exhaustion loop can step into other containers, or fail to parse any value, we can't set a carried state until we're done exhausting the contents and (from the perspective of the step_out) are back in control of what's being parsed.

In the case of `ContainerDelimiter`, if we set it before reading, and an error occurs while reading a value, the next step into `step_out` would cause the reader to start looking for a parent container's delimiter. If we set it to any other state, then another instance of `step_out` could clobber that state with whatever it was doing at the time.

I've added a unit test for your example and will add some more, and push them up when they're done.